[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148590

This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately are related):
1. When async_op=False, we directly launch the collective on "current" stream, instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves two event syncs (which have overhead in case of HIP) and one pybind when we call `work.wait()` in distributed_c10d.py on behalf of user.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with @cenzhaometa who wants to remove the event sync overhead.

Cc: @ngimel @awgu @Aidyn-A @skyw @wconstab @leonardo0lyj

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

@diff-train-skip-merge

Differential Revision: [D71652868](https://our.internmc.facebook.com/intern/diff/D71652868)

kwen2501 committed:
[ptd][nccl] use current-stream as nccl-stream under async=False mode (#147820)

Summary:

PTD current workflow:
- PTD creates its own dedicated `ncclStream` for comm operation
- it will first add a dependency on current-stream (typically the compute stream) to ensure tensors are ready before invoking collective

such stream synchronization become expensive in Inference world (cpu overhead: 70us vs GPU kernel time: 160us).

This diff:
- async=False [default], will use current-stream as nccl-stream and avoid the stream-sync overhead
- async=True, will retain existing logic: create new nccl-stream, let it wait on current-stream to ensure tensors are ready
- pass down async from c10d down to NCCL-PG

this helps shave off 50% CPU overhead **(70us -> 35us)**, which reduce total CPU/GPU from **230us to 195us by 15%**

Differential Revision: D70135605

[PGNCCL] Make avoid-record-stream default

[c10d] Add asyncOp argument to Ops

Change python side wait

Pass asyncOp at ProcessGroup level

Watchdog unstashing tensors as a safety net

lint

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately have to be atomic):
1. When async_op=False, we directly launch the collective on "current" stream instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves an event sync and one pybind during the unnecessary `work.wait()` called by distributed_c10d.py.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

taozhiwei commented:
https://github.com/pytorch/pytorch/pull/148553
I mentioned a similar one

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately have to be atomic):
1. When async_op=False, we directly launch the collective on "current" stream instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves an event sync and one pybind during the unnecessary `work.wait()` called by distributed_c10d.py.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately have to be atomic):
1. When async_op=False, we directly launch the collective on "current" stream instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves an event sync and one pybind during the unnecessary `work.wait()` called by distributed_c10d.py.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@albanD Would appreciate your help:
> The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not

We are adding `asyncOp` argument to the ops. 
Thanks!

taozhiwei commented:
> Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
> 
> * **->** [[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely #148590](https://github.com/pytorch/pytorch/pull/148590)
> 
> This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately have to be atomic):
> 
> 1. When async_op=False, we directly launch the collective on "current" stream instead of a trampoline stream and join back.
> 
> * Resolves [[RFE][Distributed][NCCL] A feature request for stream management API in PG NCCL #147729](https://github.com/pytorch/pytorch/issues/147729)
> * Resolves [[Feature][c10d] Allow flexible `cudaStreamWaitEvent` in PGNCCL #146881](https://github.com/pytorch/pytorch/issues/146881)
> * Also saves an event sync and one pybind during the unnecessary `work.wait()` called by distributed_c10d.py.
> 
> 2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
> 
> * Resolves [[FSDP2] The evil `record_stream` in c10d causes FSDP2 to over-allocate GPU memory #147168](https://github.com/pytorch/pytorch/issues/147168)
> 
> 3. Remove tensor life management when async_op=False; only use it when async_op=True.
> 4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
> 5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.
> 
> Joint work with @cenzhaometa who wants to remove the event sync overhead.
> 
> Cc: @ngimel @awgu @Aidyn-A @skyw @wconstab @leonardo0lyj
> 
> cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

In cuda graph mode，watchdogHandler is disable. when  async_op=True，and user not calling `work.wait()`, Can't input be released?

kwen2501 commented:
1. Users are in general expected to call `work.wait()` in async mode, esp when they are CUDA Graphing. Failure to do so can results in holding of the tensors. That's the side effect of making avoid-record the default. We are just using watchdog to mitigate that side effect.
2. The change to use current stream has some other reasons than tensor lifetime management, such as reducing stream sync overhead.

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately have to be atomic):
1. When async_op=False, we directly launch the collective on "current" stream instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves an event sync and one pybind during the unnecessary `work.wait()` called by distributed_c10d.py.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

taozhiwei commented:
> 1. Users are in general expected to call `work.wait()` in async mode, esp when they are CUDA Graphing. Failure to do so can results in holding of the tensors. That's the side effect of making avoid-record the default. We are just using watchdog to mitigate that side effect.
> 2. The change to use current stream has some other reasons than tensor lifetime management, such as reducing stream sync overhead.

 should be able to add a check in A `syncStream` If it's the same stream, don't call record and block anymore?

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately have to be atomic):
1. When async_op=False, we directly launch the collective on "current" stream instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves an event sync and one pybind during the unnecessary `work.wait()` called by distributed_c10d.py.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

eqy commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -711,14 +718,25 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
+void ProcessGroupNCCL::WorkNCCL::stashTensors(
+    std::vector<at::Tensor>& tensors) {
+  std::lock_guard<std::mutex> lock(stashMutex_);
+  stashed_for_allocator_safety_->insert(
```
Do we ever expect to stash tensors more than once (e.g., do multiple inserts), or realistically does this only happen once (so `stashTensors` could take ownership of the vector directly)?

eqy commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3643,13 +3639,8 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
-  }
-
+  // Therefore, we warn and fall back to the typical recordStream logic.
+  // TODO( kwen2501 ): revisit this when we have a better solution.
```
I can work on a follow-up with the solution discussed (do `wait()` in destruction if stashed tensors are alive)

eqy replied:
Actually looking at the existing code 
```
  // Both send tensor and recv tensor are created on a worker stream and used
  // in different ncclStreams.  Hence, both must record the ncclStream to
  // prevent being freed before the collective finishes.
  //
  // See [Sync Streams].
  c10::cuda::CUDACachingAllocator::recordStream(
      tensor.storage().data_ptr(), ncclStream);
 ```
 I'm wondering if this should only be done in the non-batched case. In the batched case and we have `coalescing_state_ == true`, is this redundant as either the `recordStream`/stashing would happen in `endCoalescing`, right?

kwen2501 replied:
The `pointToPoint` function is basically `isend` and `irecv`, so it is always async. So here we should do what we do for async -- stash the input/output tensor onto CPU stage. Regardless of single P2P or batch P2P.

Re `endCoalescing`, I don't think it does anything related to stash/unstash. The unstash for P2P happens in `work.wait` written by user, applicable to both single P2P and for batched P2P.

Let me know if there is anything unclear :)

Aidyn-A reviewed (approved):
Looks alright to me :+1:

fduwjj commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -363,6 +363,8 @@ class TestDebugInfoWriter : public c10d::DebugInfoWriter {
 };
 
 TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNoHeartbeat) {
+  // Note (kwen2501) 03/07/2025
```
You need to use TODO not note... I will revisit this test in the next round of BE week.

fduwjj commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -19,35 +19,35 @@ TORCH_LIBRARY(c10d, m) {
   m.def(
       "broadcast_(Tensor[] tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, int root_rank, int root_tensor, bool asyncOp, int timeout) -> (Tensor[], __torch__.torch.classes.c10d.Work)");
   m.def(
-      "allreduce_(Tensor[] tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, __torch__.torch.classes.c10d.ReduceOp reduce_op, Tensor? sparse_indices, int timeout) -> (Tensor[], __torch__.torch.classes.c10d.Work)");
```
interesting, is this definition of pybind or? For my learning purpose.

kwen2501 replied:
That's definition of a torch op.

fduwjj replied:
Got it, thanks!

albanD commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -19,35 +19,35 @@ TORCH_LIBRARY(c10d, m) {
       "broadcast_(Tensor[] tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, int root_rank, int root_tensor, bool asyncOp, int timeout) -> (Tensor[], __torch__.torch.classes.c10d.Work)");
   m.def(
-      "allreduce_(Tensor[] tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, __torch__.torch.classes.c10d.ReduceOp reduce_op, Tensor? sparse_indices, int timeout) -> (Tensor[], __torch__.torch.classes.c10d.Work)");
+      "allreduce_(Tensor[] tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, __torch__.torch.classes.c10d.ReduceOp reduce_op, Tensor? sparse_indices, bool asyncOp, int timeout) -> (Tensor[], __torch__.torch.classes.c10d.Work)");
```
These are indeed BC-breaking changes. Anyone calling these ops would have to change their callside. Isn't that going to be very disruptive?

fduwjj reviewed (approved):
Thanks for your time for offline explanation. Overall looks good to me and It would be good that we also import into fbcode and check internal CI since this is a big change.

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately are related):
1. When async_op=False, we directly launch the collective on "current" stream, instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves two event syncs (which have overhead in case of HIP) and one pybind when we call `work.wait()` in distributed_c10d.py on behalf of user.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately are related):
1. When async_op=False, we directly launch the collective on "current" stream, instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves two event syncs (which have overhead in case of HIP) and one pybind when we call `work.wait()` in distributed_c10d.py on behalf of user.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70835197).

kwen2501 commented:
@pytorchbot merge -i

janeyx99 commented:
Hi, this was a BC breaking change and is breaking the bc compat test. Reverting as a result:

@pytorchbot revert -c ignored-signal -m "PR break backward compat test"

janeyx99 commented:
@pytorchbot revert -c ignoredsignal -m "PR break backward compat test"

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately are related):
1. When async_op=False, we directly launch the collective on "current" stream, instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves two event syncs (which have overhead in case of HIP) and one pybind when we call `work.wait()` in distributed_c10d.py on behalf of user.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D70835197](https://our.internmc.facebook.com/intern/diff/D70835197)

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge -f "Added c10d::.* to BC allowlist to work around BC check; nothing else changed"

kwen2501 commented:
Note that this PR may break internal tests due to BC break, a forward fix for internal tests is here: D70854363

ZainRizvi commented:
@pytrochbot revert -c ghfirst -m "Breaking internally, see D70873275. Discussed reverting this with Ke. To validate your fixes internally, you can follow the instructions here: https://fburl.com/fixing-ghfirst-reverts"

ZainRizvi commented:
@pytorchmergebot  revert -c ghfirst -m "Breaking internally, see D70873275. Discussed reverting this with Ke. To validate your fixes internally, you can follow the instructions here: https://fburl.com/fixing-ghfirst-reverts"

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70937982).

taozhiwei commented:
I think cannot be called unstashTensors() in `synchronizeStream()` because ncclEndEvent_->block(currentStream) are issued asynchronously. the completion of the ncclEndEvent_->block(currentStream) call does not mean that the communication has ended, so the tensor cannot be released, so unstashTensors() can only be called in watchdogHandler?

kwen2501 commented:
@taozhiwei CUDACachingAllocator can run ahead of GPU, just like CPU can record events ahead of GPU completion.

ZainRizvi commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -m "reverted internally, see D71292427" -c ghfirst

Chao1Han commented:
Hi @kwen2501, May I know if this feature will be relanded soon or if modifications are required?

kwen2501 commented:
@Chao1Han @CaoE Thanks for your interest. We are investigating some internal issues. Will hopefully have some signals next week.

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately are related):
1. When async_op=False, we directly launch the collective on "current" stream, instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves two event syncs (which have overhead in case of HIP) and one pybind when we call `work.wait()` in distributed_c10d.py on behalf of user.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71652868).

kwen2501 committed:
Update on "[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely"


This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately are related):
1. When async_op=False, we directly launch the collective on "current" stream, instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves two event syncs (which have overhead in case of HIP) and one pybind when we call `work.wait()` in distributed_c10d.py on behalf of user.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with cenzhaometa who wants to remove the event sync overhead.

Cc: ngimel awgu Aidyn-A skyw wconstab leonardo0lyj

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D71652868](https://our.internmc.facebook.com/intern/diff/D71652868)

[ghstack-poisoned]

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71652868).

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
> ## Merge failed
> **Reason**: 1 mandatory check(s) failed. The first few are:
> 
> * [pull / linux-jammy-py3.9-gcc11 / test (backwards_compat, 1, 1, ephemeral.linux.2xlarge)](https://github.com/pytorch/pytorch/actions/runs/14134741815/job/39604803263)
> 
> Dig deeper by [viewing the failures on hud](https://hud.pytorch.org/pytorch/pytorch/commit/e933dfb5efa451be939f4e281ff33f5be85463ab)
> 
> Details for Dev Infra team

I don't know where this error came from as this PR does not touch related code and the first landing does not have this issue. Am going to ignore it. Can fix the issue in a different PR.

kwen2501 commented:
@pytorchbot merge -i

atalman commented:
@pytorchbot merge -f "already landed internally"
